File: Antlr/src/main/scala/io/github/chaosunity/casc/parsing/expression/FunctionCall.scala
package io.github.chaosunity.casc.parsing.expression

import io.github.chaosunity.casc.parsing.`type`.Type
import io.github.chaosunity.casc.parsing.scope.FunctionSignature

import java.util

class FunctionCall(val signature: FunctionSignature,
                   val arguments: util.List[Expression],
                   private val _owner: Expression,
                   _negative: Boolean) extends Call {
    override def `type`: Type = signature.returnType

    override def identifier: String = signature.name

    override def negative: Boolean = _negative

    def functionName: String = signature.name

    def owner: Expression = _owner

    def ownerType: Type = _owner.`type`
}

File: src/main/kotlin/io/github/chaosunity/casc/parsing/node/expression/FunctionCall.kt
package io.github.chaosunity.casc.parsing.node.expression

import io.github.chaosunity.casc.parsing.scope.FunctionSignature
import io.github.chaosunity.casc.parsing.type.Type

data class FunctionCall(
    val signature: FunctionSignature,
    override val arguments: List<Argument>,
    val owner: Expression<*>,
    val static: Boolean
) : Call<FunctionCall> {
    constructor(signature: FunctionSignature, arguments: List<Argument>, ownerType: Type, static: Boolean) : this(
        signature,
        arguments,
        EmptyExpression(ownerType),
        static
    )

    override val type: Type = signature.returnType
    override val identifier: String = signature.name
}